Makefile.docker

```diff
@@ -16,70 +16,48 @@
 # under the License.
 
 # build docker compose images:
+# $ make -f Makefile.docker build-cpp
+# To run the test suite
 # $ make -f Makefile.docker cpp
-# run the built image:
-# $ make -f Makefile.docker run cpp
 
-.PHONY: clean run cpp cpp-alpine go js java rust r
+LANGUAGES = cpp cpp-alpine cpp-cmake32 c_glib go java js python python-alpine rust r
+MISC = lint iwyu clang-format docs pandas-master
+SERVERS = dask hdfs-integration spark-integration
+
+# declare images dependencies
+DEPENDS_ON_CPP = c_glib python r
+DEPENDS_ON_CPP_ALPINE = python-alpine
+DEPENDS_ON_PYTHON = lint docs dask hdfs-integration spark-integration
+DEPENDS_ON_LINT = iwyu clang-format
+
+SERVICES = $(LANGUAGES) $(MISC) $(SERVERS)
+.PHONY: clean build-% run-% $(SERVICES)
 
 DC := docker-compose
 
 clean:
 	$(DC) down -v
 
-run:
-	$(DC) run --rm $(filter-out $@,$(MAKECMDGOALS))
-
-go:
-	$(DC) build go
-
-js:
-	$(DC) build js
-
-java:
-	$(DC) build java
-
-rust:
-	$(DC) build rust
-
-cpp:
-	$(DC) build cpp
-
-cpp-alpine:
-	$(DC) build cpp-alpine
-
-cpp-cmake32:
-	$(DC) build cpp-cmake32
-
-c_glib: cpp
-	$(DC) build c_glib
-
-r: cpp
-	$(DC) build r
-
-python: cpp
-	$(DC) build python
-
-python-alpine: cpp-alpine
-	$(DC) build python-alpine
-
-lint: python
-	$(DC) build lint
-
-iwyu: lint
-
-clang-format: lint
-
-docs: python
-
-dask: python
-	$(DC) build dask
-
-hdfs: python
-	$(DC) build hdfs-integration
+# Default build target if no dependencies
+build-%:
+	$(DC) build $(subst build-,,$@)
+
+# The following targets create the dependencies of the form `build-X: build-Y`
+$(addprefix build-,$(DEPENDS_ON_CPP)): build-cpp
```

Comment:
This is pretty fancy. While the other way is more verbose, it is a bit more readable. Thoughts?

Reply:
I simplified it to make stuff more explicit and less string manipulations.